File: src/net/auoeke/cin/element/MapElement.java
package net.auoeke.cin.element;

import java.util.HashMap;
import java.util.Map;

public class MapElement extends HashMap<String, Element> implements Element {
    public MapElement() {}

    public MapElement(int initialCapacity, float loadFactor) {
        super(initialCapacity, loadFactor);
    }

    public MapElement(int initialCapacity) {
        super(initialCapacity);
    }

    public MapElement(Map<? extends String, ? extends Element> source) {
        super(source);
    }

    @Override public Type type() {
        return Type.MAP;
    }

    @Override public MapElement clone() {
        return new MapElement(this);
    }

    @Override public String toString() {
        return switch (this.size()) {
            case 0 -> "{}";
            case 1 -> '{' + this.format(this.entrySet().iterator().next().getValue()) + '}';
            default -> {
                var builder = new StringBuilder("{\n");
                this.forEach((key, value) -> builder.append("    ").append(this.format(value).replaceAll("\n", "\n    ")).append('\n'));

                yield builder.append('}').toString();
            }
        };
    }

    private String format(Element value) {
        return switch (value.type()) {
            case ARRAY, MAP -> "$key $value";
            default -> "$key = $value";
        };
    }
}

File: src/net/auoeke/cin/element/MapElement.kt
package net.auoeke.cin.element

class MapElement : HashMap<String, Element>(), Element {
    override val type: ElementType get() = ElementType.MAP

    override fun clone(): MapElement = MapElement().also {it += this}

    override fun toString(): String = when (size) {
        0 -> "{}"
        1 -> entries.first().let {
            "{${format(it)}}"
        }
        else -> buildString {
            append("{\n")

            this@MapElement.forEach {entry ->
                append(format(entry).prependIndent("    "))
                append('\n')
            }

            append('}')
        }
    }

    private fun format(entry: Map.Entry<String, Element>): String = entry.let {(key, value) ->
        when (value.type) {
            ElementType.ARRAY, ElementType.MAP -> "$key $value"
            else -> "$key = $value"
        }
    }
}
